### PR TITLE
added if_scalar_cast_to_list function to schedule.filter()

### PR DESCRIPTION
**Pull request by NuoWenLei**
### Summary

extension fix of #3646 
### Details and comments
Tested with scalar values of each parameter.

### CODE REVIEW

File: qiskit/pulse/schedule.py [hunk 1]
```diff
@@ -324,17 +324,28 @@ def _construct_filter(self, *filter_funcs: List[Callable],
             time_ranges: For example, [(0, 5), (6, 10)]
             intervals: For example, [Interval(0, 5), Interval(6, 10)]
         """
-        def only_channels(channels: Set[Channel]) -> Callable:
+
+        def if_scalar_cast_to_list(to_list):
+            return_ls = []
+            if str(type(to_list)) != "<class 'list'>":
```
Comment (taalexander) [inline qiskit/pulse/schedule.py]: This could be
```python
try:
    to_list = iter(to_list)
except TypeError:
    to_list = [to_list]

**Comment by NuoWenLei** [comment]
Hello @taalexander , sorry but I don't understand what you mean by tests? Also, I've changed the if_scalar_cast_to_list function as you've suggested.

**Comment by lcapelluto** [comment]
Hi @NuoWenLei ! Thank you so much for your contribution -- this looks great! Please test the new functionality, and fix the merge conflicts. Then we can get it accepted :D

**Comment by lcapelluto** [comment]
Hi @NuoWenLei, any update?

**Comment by NuoWenLei** [comment]
I've just resolved the merge conflicts

**Comment by taalexander** [comment]
Hi @NuoWenLei, there are now new conflicts. My apologies for the delay, I missed this in my roundup of open PRs.

**Comment by taalexander** [comment]
It seems we have one more conflict here that is causing tests to fail. An itertools import is now missing.

**Comment by NuoWenLei** [comment]
Sorry, it seems that the version of schedule.py in my local repository did not have some of the imports. I'm a little perplexed about why this happened, but I will fix it now.

File: qiskit/pulse/schedule.py [hunk 2]
```diff
@@ -334,49 +334,54 @@ def _apply_filter(self, filter_func: Callable, new_sched_name: str) -> 'Schedule
...
-
                 Args:
                     time_inst (Tuple[int, Instruction]): Time
                 """
                 return any([chan in channels for chan in time_inst[1].channels])
             return channel_filter
 
-        def only_instruction_types(types: Iterable[abc.ABCMeta]) -> Callable:
-            def instruction_filter(time_inst) -> bool:
+        def only_instruction_types(types: Optional[Iterable[abc.ABCMeta]] = None) -> Callable:
```
Comment (taalexander) [inline qiskit/pulse/schedule.py:371]:
```suggestion
        def only_instruction_types(types: Optional[Iterable['Instruction']] = None) -> Callable:
```

File: qiskit/pulse/schedule.py [hunk 3]
```diff
@@ -334,49 +334,54 @@ def _apply_filter(self, filter_func: Callable, new_sched_name: str) -> 'Schedule
 
     def _construct_filter(self, *filter_funcs: List[Callable],
                           channels: Optional[Iterable[Channel]] = None,
-                          instruction_types=None,
+                          instruction_types: Optional[Iterable[Type['Instruction']]] = None,
```
Comment (taalexander) [inline qiskit/pulse/schedule.py:337]:
```suggestion
                          instruction_types: Optional[Iterable['Instruction']] = None,
```

**Comment by taalexander** [comment]
Just a couple of small changes and then it is good to go 😄.

**Comment by NuoWenLei** [comment]
What is the best way to resolve the branch conflicts?

**Comment by taalexander** [comment]
> What is the best way to resolve the branch conflicts?

A merge is fine. All changes will be squashed before merging into master.

**Comment by NuoWenLei** [comment]
Sorry, the indentation was off.

File: qiskit/pulse/schedule.py [hunk 4]
```diff
@@ -327,80 +308,77 @@ def _apply_filter(self, filter_func: Callable, new_sched_name: str) -> 'Schedule
...
+            intervals: For example, [Interval(0, 5), Interval(6, 10)] or Interval(0, 5)
         """
-        def only_channels(channels: Set[Channel]) -> Callable:
-            def channel_filter(time_inst) -> bool:
-                """Filter channel.
 
-                Args:
-                    time_inst (Tuple[int, Instruction]): Time
-                """
+        def if_scalar_cast_to_list(to_list):
```
Comment (NuoWenLei) [inline qiskit/pulse/schedule.py:331]: Hi @taalexander,
I went through most of the errors in the checks, and all that I saw pointed to an indentation error on this line. The error is: IndentationError: unindent does not match any outer indentation level

I'm sorry for such a simple problem. I tried to re-upload, edit from a previous edition of the file, but none seem to satisfy the standard. Do you know what's going on here?

Comment (taalexander) [reply qiskit/pulse/schedule.py:331]: Is it possible that line 314 has ane extra space?

Comment (NuoWenLei) [reply qiskit/pulse/schedule.py:330]: Indentation is by 4 spaces, right? I'm pretty sure that the lines are indented correctly. This is quite interesting. Maybe I could do a regex search to see if I might have mixed tabs and spaces? What's the syntax for it?

**Comment by taalexander** [comment]
The line 314 has 9 space before the docstring.

**Comment by NuoWenLei** [comment]
Sorry for such an embarrassing miss. That's quite frustrating.

**Comment by NuoWenLei** [comment]
I'll fix the lint problems.

**Comment by taalexander** [comment]
> Sorry for such an embarrassing miss. That's quite frustrating.

That's what tests, CI and reviews are for 😄 . No worries at all.

> I'll fix the lint problems. There is a type error while using Interval(start, stop). Could it be that I did not import Interval from .timeslots?

Timeslots we're changed for performance reasons. I think you can have a look at the current filter implementation to update.

**Comment by NuoWenLei** [comment]
Is there a way to access the tests of previous commits? Interval is defined in the file as:
Interval = Tuple[int, int]
So I'd like to check what the problem was during the second to last test.

**Comment by taalexander** [comment]
I'm not certain. You could always checkout that commit locally and run the tests. Or submit a dummy PR against Qiskit to have CI do it.

**Comment by NuoWenLei** [comment]
I just checked it out and the error was a type error:

TypeError: Type Tuple cannot be instantiated; use tuple() instead

I looked up this problem in more specificity.

On Github, there mentioned that the instantiation of generic versions of built-in classes are not allowed, and the generic versions of Tuple could also not be instantiated due to other reasons (something about it not being a regular class). Do you think this is what is happening here?
(File mentioning this is pep-0484.txt)

**Comment by taalexander** [comment]
> I just checked it out and the error was a type error:
> 
> TypeError: Type Tuple cannot be instantiated; use tuple() instead
> 
> I looked up this problem in more specificity.
> 
> On Github, there mentioned that the instantiation of generic versions of built-in classes are not allowed, and the generic versions of Tuple could also not be instantiated due to other reasons (something about it not being a regular class). Do you think this is what is happening here?
> (File mentioning this is pep-0484.txt)

Where are you seeing this error? I don't see it being raised in CI?

**Comment by NuoWenLei** [comment]
Sorry, that was a problem I had before, but now I think the code functionality is good. The problems left I think are just format problems, I will fix promptly.

**Comment by NuoWenLei** [comment]
The only error left is that there is some kind of IndentationError. Sorry, I don't really understand this error:
/home/vsts/work/1/s/qiskit/pulse/schedule.py:docstring of qiskit.pulse.Schedule:6:Unexpected indentation.

**Comment by NuoWenLei** [comment]
@taalexander what does `r"""` mean in the code? Could that be the error for the unexpected indentation?

**Comment by 1ucian0** [comment]
ping @NuoWenLei ? The status of this PR is in "changes requested" for some time. This is also a conflict to be resolved. Are you still working on this?

**Comment by NuoWenLei** [comment]
Sorry, I had not been working on it after I got stuck on the IndentationError. I just committed the resolved conflicts, but I likely will not be working on this after.

**Comment by NuoWenLei** [comment]
The last error now should be the IndentationError again

**Review by taalexander — APPROVED** [review]

**Pull request merged**